gapis/api/vulkan/read_framebuffer.go

```diff
@@ -1007,129 +1060,200 @@ func postImageData(ctx context.Context,
+			NewVkCommandBufferᶜᵖ(commandBuffers.Ptr()), // pCommandBuffers
+			0, // signalSemaphoreCount
+			0, // pSignalSemaphores
+		)
+		submitInfoData := MustAllocData(ctx, s, submitInfo)
+		writeEach(ctx, out,
+			cb.VkQueueSubmit(
+				vkQueue,
+				1,
+				submitInfoData.Ptr(),
+				VkFence(0),
+				VkResult_VK_SUCCESS,
+			).AddRead(
+				submitInfoData.Data(),
+			).AddRead(
+				commandBuffers.Data(),
+			),
+			cb.VkDeviceWaitIdle(vkDevice, VkResult_VK_SUCCESS),
+		)
+	}
 
-	// Dump the buffer data to host
-	writeEach(ctx, out,
-		cb.VkMapMemory(
-			vkDevice,
-			bufferMemoryID,
-			VkDeviceSize(0),
-			VkDeviceSize(bufferSize),
-			VkMemoryMapFlags(0),
-			mappedPointer.Ptr(),
-			VkResult_VK_SUCCESS,
-		).AddWrite(mappedPointer.Data()),
-		cb.VkInvalidateMappedMemoryRanges(
-			vkDevice,
-			1,
-			mappedMemoryRangeData.Ptr(),
-			VkResult_VK_SUCCESS,
-		).AddRead(mappedMemoryRangeData.Data()),
-	)
+	t.pendingReads = append(t.pendingReads, pendingRead{
+		device:             vkDevice,
+		buffer:             bufferID,
+		bufferMemory:       bufferMemoryID,
+		bufferSize:         bufferSize,
+		commandPool:        commandPoolID,
+		stagingImage:       stagingImageID,
+		stagingMemory:      stagingImageMemoryID,
+		resolveImage:       resolveImageID,
+		resolveImageMemory: resolveImageMemoryID,
+		aspect:             aspect,
+		format:             vkFormat,
+		imageFormat:        formatOfImgRes,
+		requestWidth:       requestWidth,
+		requestHeight:      requestHeight,
+		res:                res,
+	})
+}
 
-	// Add post command
-	writeEach(ctx, out,
-		cb.Custom(func(ctx context.Context, s *api.GlobalState, b *builder.Builder) error {
-			b.Post(value.ObservedPointer(at.Address()), uint64(bufferSize), func(r binary.Reader, err error) {
-				var bytes []byte
-				if err == nil {
-					bytes = make([]byte, bufferSize)
-					r.Data(bytes)
-					r.Error()
-
-					// For the depth aspect of VK_FORMAT_X8_D24_UNORM_PACK32 and
-					// VK_FORMAT_D24_UNORM_S8_UINT format, we need to strip the
-					// undefined value in the MSB byte.
-					if aspect == VkImageAspectFlagBits_VK_IMAGE_ASPECT_DEPTH_BIT && (vkFormat == VkFormat_VK_FORMAT_X8_D24_UNORM_PACK32 || vkFormat == VkFormat_VK_FORMAT_D24_UNORM_S8_UINT) {
-						inBufSize := 4
-						inImgSize := 3
-						count := len(bytes) / inBufSize
-						for i := 0; i < count; i++ {
-							copy(bytes[i*inImgSize:(i+1)*inImgSize], bytes[i*inBufSize:(i+1)*inBufSize])
+func (t *readFramebuffer) Flush(ctx context.Context, out transform.Writer) {
+	t.FlushPending(ctx, out)
+}
+
+func (t *readFramebuffer) FlushPending(ctx context.Context, out transform.Writer) {
+	s := out.State()
+	cb := CommandBuilder{Thread: 0, Arena: s.Arena}
+	a := s.Arena // TODO: Use a temporary arena?
+	for i := range t.pendingReads {
+		// DO NOT BE TEMPTED TO TURN THIS INTO
+		// for _, f := range t.pendingReads.
+		// It will not do what you want.
+		// There is a lambda capture down there, it will not do what
+		// you expect.
```

Comment:
This, the indentation becoming pretty high, and global readability: would it be possible to replace the lambda captures and have the lambdas defined as proper named functions?